### DISCUSSION TITLE
Is there any way to track field changes preserving the stack trace?

### DISCUSSION THREAD

**Discussion by shinobu-uwu** [pmndrs/zustand#2733] [Q&A] [answered]
I'm debugging an issue that I'm facing with zustand where out of nowhere two fields of my store go to `undefined`. I am using the persist middleware and already checked that it is being commited to the storage with them being `undefined`.

What is weird is that I don't set them to `undefined` at any moment, and it's only being persisted as `undefined` on unrelated operations, so my guess is that they are silently being set, for some inexplicable reason, and when I do another set operation it's commited to the storage and then saved as `undefined`.

So back to my question, is there anyway to track these values? Subscribe didn't work well, it doesn't preserve the stack trace and I think it didn't account for the "silent change". Unfortunately I cannot share my code, so feel free to ask any further questions

**Comment by dbritto-dev** [comment, marked as answer]
@shinobu-uwu sure, you can create a custom middleware for logging and track your changes or you can use subscribe

**Reply by shinobu-uwu** [reply]
I used the log middleware from [this discussion](https://github.com/pmndrs/zustand/discussions/2704) and I think I figured out why it's not working.

My guess is that since I'm developing a chrome extension, the two fields in question are being set in the service worker context, which runs isolated from the main page context, so when I set another field from that same store in the main page it looks like zustand overwrites them with the initial value, which is indeed `undefined`.

The thing is, this shouldn't be happening because the store is hydrated when the page loads, and those two fields are always being set when the page is closed, which should have the correct value for those two fields...

What is weirder is that this only happens when I reload the extension, which reloads the service worker.

I'm not as lost as before but still confused about all of this, I'll report back when I find something else that could be the root of this

**Reply by dbritto-dev** [reply]
@shinobu-uwu are you checking that the store is hydrated? btw, here you go -> https://stackblitz.com/edit/nextjs-and-zustand-persist?file=app%2F_components%2Fhome-page.tsx

**Reply by shinobu-uwu** [reply]
yes I'm waiting for it to be hydrated

**Reply by shinobu-uwu** [reply]
I think I got auth to work, however I'm still facing the issue of the service worker store not being synchronized with the main page, is there anyway to manually trigger rehydration?

**Reply by shinobu-uwu** [reply]
I ask that because if I refresh the service worker then the store is correct, so I think rehydrating would fix it

**Reply by dbritto-dev** [reply]
@shinobu-uwu yeah, there's a way to manually rehydrate, actually in those cases you need to disable auto hydration and do it manually in `useEffect`/`useLayoutEffect`
